test/promql/tests/data_plane/gitops_sync_test.yaml

```diff
@@ -0,0 +1,101 @@
+        values: '1 1 1 1 1 1'
+      - series: 'argocd_app_info{sync_status="Unknown",namespace="gitops-service-argocd"}'
+        values: '1 1 1 1 1 1'
+
+    alert_rule_test:
+      - eval_time: 6m
+        alertname: GitOpsApplicationSyncErrors
+        exp_alerts:
+          - exp_labels:
+              severity: warning
+              slo: true
+            exp_annotations:
+              summary: >-
+                Less than 95% of GitOps applications are in synced state
+              description: >-
+                The percentage total of all Argo CD applications that are in Synced state is less than 95%.
+              runbook_url: https://gitlab.cee.redhat.com/rhtap/docs/sop/-/blob/main/gitops/deploy-from-git-to-k8s.md
+
+  - interval: 1m
+    input_series:
+
+      # GitOps Service experienced less than 95% synced rate in applications, should alert
+      - series: 'argocd_app_info{sync_status="Synced",namespace="gitops-service-argocd"}'
+        values: '1 1 1 1 1 1'
+      - series: 'argocd_app_info{sync_status="OutOfSync",namespace="gitops-service-argocd"}'
+        values: '1 1 1 1 1 1'
+      - series: 'argocd_app_info{sync_status="Unknown",namespace="gitops-service-argocd"}'
+        values: '_ _ _ _ _ _'
+
+    alert_rule_test:
+      - eval_time: 6m
+        alertname: GitOpsApplicationSyncErrors
+        exp_alerts:
+          - exp_labels:
+              severity: warning
+              slo: true
+            exp_annotations:
+              summary: >-
+                Less than 95% of GitOps applications are in synced state
+              description: >-
+                The percentage total of all Argo CD applications that are in Synced state is less than 95%.
+              runbook_url: https://gitlab.cee.redhat.com/rhtap/docs/sop/-/blob/main/gitops/deploy-from-git-to-k8s.md
+
+  - interval: 1m
+    input_series:
+
+      # GitOps Service experienced greater than 95% synced rate for applications, should not alert
+      - series: 'argocd_app_info{sync_status="Synced",namespace="gitops-service-argocd"}'
+        values: '1 1 1 1 1 1'
+      - series: 'argocd_app_info{sync_status="OutOfSync",namespace="gitops-service-argocd"}'
+        values: '_ _ _ _ _ _'
+      - series: 'argocd_app_info{sync_status="Unknown",namespace="gitops-service-argocd"}'
+        values: '_ _ _ _ _ _'
+
+    alert_rule_test:
+      - eval_time: 6m
+        alertname: GitOpsApplicationSyncErrors
+
+  - interval: 1m
+    input_series:
+
+      # GitOps Service experienced greater than 95% synced rate for applications, should not alert
+      - series: 'argocd_app_info{sync_status="Synced",namespace="gitops-service-argocd"}'
+        values: '1 1 1 1 1 1'
+      - series: 'argocd_app_info{sync_status="OutOfSync",namespace="gitops-service-argocd"}'
+        values: '_ _ 1 _ 1 _'
+      - series: 'argocd_app_info{sync_status="Unknown",namespace="gitops-service-argocd"}'
+        values: '_ _ _ 1 _ 1'
```

Comment:
I think (no sure), that if you'd make those time series a little longer, then you'd get an alert.

The reason being is that, for Prometheus, the value of the 2 non-synced time series is still `1` as due to the nature of this metric, it cannot get any other value other than `1`.

The way to overcome it is to have the alert rule examine specific time windows, but there we have issues as well.
```
      - series: 'argocd_app_info{sync_status="Synced",namespace="gitops-service-argocd"}'
        values: '1 1 1 1 1 1 1 1 1 1 1'
      - series: 'argocd_app_info{sync_status="OutOfSync",namespace="gitops-service-argocd"}'
        values: '_ _ 1 _ 1 _ _ _ _ _ _'
      - series: 'argocd_app_info{sync_status="Unknown",namespace="gitops-service-argocd"}'
        values: '_ _ _ 1 _ 1 _ _ _ _ _'
```

There might be a way to do it, but there are many corner cases to consider here. 

One option that might be less fragile (I think it's an overkill) is to have a **recording rule** that will aggregate the combination of such time series, for each app, to a single metric that will provide 1 if it's synced and 0 otherwise. Then we can have an **alerting rule** utilizing that metric.

I'm not optimistic about the ability to make those computations based on the `argocd_app_info` metric (I'll be glad to be proven wrong).

Reply:
@yftacherzog , all your comments are enlightening.

> I think (not sure), that if you'd make those time series a little longer, then you'd get an alert.

It did not. I added additional tests to show how to get an alert.

> The reason being is that, for Prometheus, the value of the 2 non-synced time series is still 1 as due to the nature of this metric, it cannot get any other value other than 1.

In the tests, I made them distinct by name, so each series should have a value 1 or not. This seems to be confirmed by looking at the OpenShift console Metrics page.

I checked in another commit and included descriptive comments for each test.

Reply:
It behaves a little different than what I thought, but still there's a delicate point here.

Contrary to what I suggested above, the latest timestamp being considered needs to be within the window, but Prometheus will assume it to be unchanged until it gets a new value (which it never will for this metric). It means that the 5m window we give is only accumulating noise rather than eliminating it.

The test below does raise the alert (I don't think that's the rule's desired behavior).
It has a single app that was out of sync for 1 minute and in sync for the rest of the time window. The `1` value provided when it was out of sync is still being considered as long as it's still within the rule's window (5m), and while we should be 100% in sync, we're actually only 50% in sync and we get an alert. 

```
      - series: 'argocd_app_info{sync_status="Synced",name="gitopsdepl-001",namespace="gitops-service-argocd"}'
        values: ' 1 _ 1 1 1 1'
      - series: 'argocd_app_info{sync_status="OutOfSync",name="gitopsdepl-001",namespace="gitops-service-argocd"}'
        values: ' _ 1 _ _ _ _'

    alert_rule_test:
      - eval_time: 6m
        alertname: GitOpsApplicationSyncErrors
```

Reply:
Maybe (really maybe) we can mitigate that with evaluating range vectors of small range (say 1m) over the 5 minutes period. This should allow us to have the alert wait for few minutes before firing while considering short time windows so the issue we see above is somewhat diminished a little.

Something like this (needs to be combined with my suggestion for eliminating the no-synced scenario:
```
      expr: |
        sum(
          last_over_time(
            argocd_app_info{sync_status="Synced", namespace="gitops-service-argocd"}[1m]
          )
        )
        /
        sum(
          last_over_time(
            argocd_app_info{namespace="gitops-service-argocd"}[1m]
          )
        )
        < 0.95
      for: 5m
```

Reply:
I tried the above but it focusses on a  very narrow window and it depends on the when the evaluation occurs.  So it can alert when the app is out of sync for that specific window, even though the app can be in sync for several intervals.

I tried `avg_over_time` over `[5m]` and `for 5m`, which expands the window.

My test cases does show that it makes sense to do this, and probably explains why I'm getting 50%, when you're expecting to get 0%.

Reply:
The only value this metric is generating is `1`, so average over time will be the same as max over time (or min over time), and for that reason, I think the use of average here is misleading.

The reason we're getting 50% is that the `argocd_app_info{sync_status="Synced"}` returns one series over that period and `argocd_app_info` returns two series over that period (the synced and the out of sync). Each of those have a value of `1`, so the sum of the synced is `1` and the sum of both synced and out of sync is `2`, and this is why we're getting `1/2`.

Reply:
I'd say again what I said at the very beginning, I don't think the `argocd_app_info` metric is a good fit for measuring this SLO. My suggestion is to either change the SLO or to think of another way to measure it.

Reply:
I can switch it back to last_over_time over [1m].  I think this is the best we can do.  Perhaps we will not have the precise or perfect expression/formula.  The measurement is actually 'arbitrary' for a given window anyway, where the application sync state can dramatically change outside of that window.  The other perspective in favour of having the 1m window is a quick 'snapshot' of the system (regardless of how old a given series value is).  Trying out all the different test scenarios showed that for those situations where the sync state is stable (not changing over a reasonable period of time), we do get the expected results to alert or not to alert.  Typically, when an app is not synced or is unknown, it will likely stay that way until there is intervention.  eg. see the test cases where I had 20 apps and 10 apps.  We can also tweak the acceptance rate of 95% and lower it if we are getting alerted too often

Reply:
ok